psh/date/date.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Phoenix-RTOS
+ *
+ * date - print or set the system date and time
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include "../psh.h"
+
+
+char *wdaylookup[] = { "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat" };
+
+char *monlookup[] = { "Jan", "Feb", "Mar", "Apr", "May", "June", "July", "Aug", "Sept", "Oct", "Nov", "Dec" };
+
+enum mode { SKIP,
+	HELP,
+	PRINT,
+	WRITE };
+
+static void psh_date_info(void)
+{
+	printf("print/set the system date and time");
+}
+
+
+static void psh_date_help(const char *prog)
+{
+	printf("Usage: %s [option] [epoch]\n", prog);
+	printf("  -h:  shows this help message\n");
+	printf("  -s:  set system time described by EPOCH (POSIX time format)\n");
+}
+
+
+static int psh_date_print(void)
+{
+	time_t t;
+	struct tm *tinfo;
+
+	time(&t);
+	tinfo = localtime(&t);
+
+	printf("%s,", wdaylookup[tinfo->tm_wday]);
+	/* `struct tm` field tm_year is years since 1900, so modulo for last 2 digits */
+	printf(" %02d %s %d", tinfo->tm_mday, monlookup[tinfo->tm_mon], (tinfo->tm_year) % 100);
+	printf(" %02d:%02d:%02d", tinfo->tm_hour, tinfo->tm_min, tinfo->tm_sec);
+	printf("\n");
+
+	return 0;
+}
+
+
+static int psh_date_set(char *timestring)
+{
+	struct timeval tv = { .tv_sec = 0, .tv_usec = 0 };
+	struct timezone tz = { .tz_dsttime = 0, .tz_minuteswest = 0 };
+
+	if (timestring == NULL) {
+		return 1;
+	}
+
+	if (strlen(timestring) == 1 && timestring[0] == '0') {
+		tv.tv_sec = 0;
+	}
+	else {
+		tv.tv_sec = strtoull(timestring, NULL, 10);
+		if (tv.tv_sec == 0) {
+			fprintf(stderr, "Uncrecognized argument \"%s\"!\n", timestring);
+			return 1;
+		}
+	}
+
+	return settimeofday(&tv, &tz);
```

Comment:
I would like to see `strerror` WHY the date couldn't be set (for example currently we may pass negative value in `tv.tv_sec` by passing large enough unsigned long long value as `time_t` is a signed type - but that for sure is tested in the appropriate tests added :P)

Reply:
is `time_t` indeed signed? According to https://github.com/phoenix-rtos/libphoenix/blob/d5b8ba661c3c829ce4bb2ec3b0a7d3d8bf9bc136/include/arch/ia32/types.h#L23
it looks like `unsigned long long int`. Am I checking it wrong? 

However - added check for negative values passed to `date -s` and added edge cases in tests.

Reply:
Yup, it's unsigned, perhaps should be signed.

Reply:
I probably wasn't specific enough, `ISO C` doesn't impose time_t type (says it might be integer or real type - signed or not), `POSIX.1-2017` specifies `time_t` to be integer type (https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_types.h.html - can be signed or not), GNU C specifies it to be signed integer type.

I would prefer code which works regards of the implementation detail of how `time_t` is defined :).

Side note: `time_t` is defined in kernel and libphoenix separately, hopefully will be fixed soon as a part of kernel headers cleanup.

Reply:
Ok, thanks!. Now I understood your worries.

As it come out we are currently storing `time_t` as `unsigned long long` and by passing negative value to `tv.tv_sev` we end up with some value which is between `max(unsigned long long)` and `max(signed long long)`. 

E.g if we execute `date -s -2` it will set to some huge value, and after 2 seconds overflow to be `0`.

Added `strerror()` read if `settimeofday()` fails. However it does not fail currently so often. But in the future it may be handy. I removed tests regarding passing negative vaues and may substitute them with ones that test __ if it works as it works__ rather than __if it works as it should__ just for regression check. But this is a topic for tests PR.